example/gmt/sphere_preprocessing.c

```diff
@@ -0,0 +1,525 @@
+  (at your option) any later version.
+
+  p4est is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with p4est; if not, write to the Free Software Foundation, Inc.,
+  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+/** \file sphere_preprocessing.c
+ *
+ * Preprocessing for the sphere model. The main function reads in a list of
+ * geodesics and splits them into segments, such that each segment is 
+ * contained entirely in a single cube face.
+ * 
+ * Usage: p4est_sphere_preprocessing <input.csv>
+ * 
+ * Here <input.csv> is a CSV file where each line
+ *    phi1,theta1,phi2,theta2
+ * represents the unique geodesic between endpoints (phi1, theta1) and
+ * (phi2, theta2). We require that the endpoints are not antipodal, as in this
+ * case there is not a unique geodesic between them.
+ * 
+ * Endpoints are given in spherical coordinates. We take the convention
+ * described here:
+ * https://en.wikipedia.org/wiki/Spherical_coordinate_system.
+ * That is, a spherical coordinate is a pair (phi, theta) where:
+ *  0 <= theta <= 180 is the polar angle
+ *  0 <= phi <= 360   is the azimuth
+ */
+
+#include "gmt_models.h"
+
+/* Convert from angular coordinates to corresponding point on cube face */
+static void
+angular_to_cube (const double angular[2], double xyz[3])
+{
+  double              inf_norm;
+  double              phi, theta;
+
+  /* Convert to radians */
+  phi = angular[0] * M_PI / 180.0;
+  theta = angular[1] * M_PI / 180.0;
+
+  xyz[0] = sin (theta) * cos (phi);
+  xyz[1] = sin (theta) * sin (phi);
+  xyz[2] = cos (theta);
+
+  inf_norm = fmax (fabs (xyz[0]), fmax (fabs (xyz[1]), fabs (xyz[2]))) * 2.0;
+  xyz[0] /= inf_norm;
+  xyz[1] /= inf_norm;
+  xyz[2] /= inf_norm;
+  return;
+}
+
+/**
+ * Which face does a given point belong to
+ *
+ * Note: ties are decided arbitrarily for the time being
+ *
+ * \param[in] xyz  cartesian coordinates on surface of the cube [-0.5,0.5]^3
+ */
+static int
+point_to_tree (const double xyz[3])
+{
+  if (fabs (xyz[0] + 0.5) < SC_EPS)
+    return 2;
+  if (fabs (xyz[0] - 0.5) < SC_EPS)
+    return 5;
+  if (fabs (xyz[1] + 0.5) < SC_EPS)
+    return 4;
+  if (fabs (xyz[1] - 0.5) < SC_EPS)
+    return 1;
+  if (fabs (xyz[2] + 0.5) < SC_EPS)
+    return 0;
+  if (fabs (xyz[2] - 0.5) < SC_EPS)
+    return 3;
+  return -1;                    /* This should not happen */
+}
+
+/**
+ * coordinate transformation from the surface of cube [-0.5,0.5]^3
+ * to tree-local coordinates. This is the inverse of 
+ * p4est_geometry_cubed_X
+ *
+ * \param[in]  xyz  cartesian coordinates in physical space
+ * \param[in]  which_tree face of cube
+ * \param[out] rst  coordinates in AMR space : [0,1]^3
+ *
+ */
+static void
```

Comment:
Would it make sense to P4EST_ASSERT the ranges of input arguments?

Reply:
Good suggestion! I just realised I actually fail these assertions due to numerical round-off issues.

I'll have a think about the best way to address this.